courseworks/week9/yushengzhao/hw9.org

```diff
@@ -0,0 +1,93 @@
+ > 1$. Let's assume $\sum_{k = 0}^{\delta} \frac{\tau (k+1)...\delta}{(\tau
+ +k)...(\tau + \delta)} = poly(\tau)$ and then multiply both sides by $(\tau ) *
+ ....*(\tau + \delta)$. We get,
+
+ \begin{equation}
+    \sum_{k=0}^{\delta} \tau * (k+1) * ... * \delta * \tau * ... * (\tau +k+1) = \tau * ...  *(\tau + \delta) * poly(\tau)
+ \end{equation}
+
+ We then compare the highest order term in $\tau$ on both sides. The l.h.s gives
+ $\tau * \delta! * \frac{\tau^{\delta -1}}{\delta!} = \tau^\delta$. And the
+ r.h.s gives $\tau^delta$ multilied by the leading order of $poly(\tau)$.
+ Therefore, the leading order of $poly(\tau)$ must be $1$. Since $\poly(\tau)$
+ does not contain $\tau^{-n}$ terms, $poly(\tau) = 1$.
+
+ Therefore, we have reached the conclusion that $\sum_{k = 0}^{\delta}
+ \frac{\tau (k+1)...\delta}{(\tau +k)...(\tau + \delta)} = 1$. Therefore,
+ $\sum_{k = 0}^{\delta} \eta(\tau - 1, k) = \eta(\tau,\delta)$.
+
+* Problem 2
+** Forward Mode Diff
+#+begin_src julia
+# the input is a vector of Duplicated type as defined in Enzyme
+# we denote abs2() as y1()
+# we denote sqrt() as y2()
+function fd_poorman_norm(x::Vector{Duplicated{<:Real}})
+    #initialization does not get derivatives
+    vtype, dtype = eltype(x).types
+    dual_nm2 = Duplicated(zero(val_type),zero(deri_type))
+    for i=1:length(x)
+        # diff_abs2(x[i]) denotes \frac{\partial y_1}{\partial x_i}
+        # note, we don't treat x[i] as independent variable
+        # i.e, we allow \frac{\partial x_i}{\partial x_i} != 1
+        dual_nm2 += Duplicated(x[i].val,x[i].dval * diff_abs2(x[i].val))
+    end
+    # similarily, we denote diff_sqrt(nm2.val) as \frac{\partial y2}{\partial y1}
+    ret = Duplicated(sqrt(dual_nm2.val), diff_sqrt(nm2.val) * dual_nm2.dval)
+    return ret
+end
+#+end_src
+
+**  Reverse Mode Diff
+#+begin_src julia
+function rd_poorman_norm(x::Vector{<:Real})
+    # start of forward pass
+    nm2 = zero(real(eltype(x)))
+    for i=1:length(x)
+        nm2 += abs2(x[i])
+    end
+
+    # nm2 -> T
+    # we need to cache nm2 value here
+    # although, I  highly doubt that
+    # you will need to explicitly cache nm2 since
+    # the variable storing it will not be modified
+
+    ret = sqrt(nm2)
+
+    # note, this is sort of pseudo code, I  don't  put
+    # the cached  part into consideration.
+    return ret, bk_poorman_norm(nm2)
+end
+
+function bk_poorman_norm(nm2::Vector{<:Real})
+
+    # start to do the backward pass
+    # this is \bar{ret}
+    ret_bar = diff_sqrt(nm2)
+
+    # nm2 <- T
+    # calculate a vector of diff
+    nm2_bar = diff_abs2.(x)
+    total_bar = ret_bar * nm2_bar
+    return total_bar
+end
+#+end_src
+** Estimate cached number
+I found it hard to understand this problem, we only have one jacobian to compute
+in order to get to the final derivative of this function. Therefore, we only
+need to cache one jacobian of dimension $1 \times length(x)$.
```

Comment:
For this problem I think in a reversed mode autodiff there should be a stack storing the data of each step delivering toghter with the variables, here is my code:
```
# this part is the forward process
function reverse_norm_forward!(x::Vector{<:Real}, stack::Vector{<:Real})
    nm2 = zero(real(eltype(x)))
	for i=1:length(x)
		push!(stack, x[i])
		nm2 += abs2(x[i])
	end
	push!(stack, nm2)
	ret = sqrt(nm2)
	return ret, stack
end

# this part is the backward process
function reverse_norm_backward!(partial_x::Vector{<:Real}, stack::Vector{<:Real})
	nm2 = pop!(stack)
	partial_x .*= 1 / (2 * sqrt(nm2))
	N = length(partial_x)
	for i in 1:N
		x = pop!(stack)
		partial_x[N + 1 - i] *= 2 * x
	end
	return partial_x, stack
end
```
In this case, it will be much more easier to couple this function with other functions to form more complex programs.
And the cache we have to use for the poorman norm function will be $N + 1$, where $N$ is the length of variable ```x```.

Reply:
Corrected